src/nvim/diff.c

```diff
@@ -1101,20 +1111,104 @@ static int diff_file_internal(diffio_T *diffio)
+  listitem_T *n;
+  typval_T *c_val;
+  typval_T *n_val;
+  int i_size;
+  long start_a;
+  long count_a;
+  long start_b;
+  long count_b;
+  int success = FAIL;
+  if (c != NULL) {
+    for (; c != NULL; c = TV_LIST_ITEM_NEXT(chunks, c)) {
+      c_val = TV_LIST_ITEM_TV(c);
+      if (c_val->v_type == VAR_LIST) {
+        i = c_val->vval.v_list;
+        i_size = tv_list_len(i);
+        if (i_size == 4) {
+          n = tv_list_first(i);
+          // Check information types.
+          for (; n != NULL; n = TV_LIST_ITEM_NEXT(i, n)) {
+            n_val = TV_LIST_ITEM_TV(n);
+            if (n_val->v_type != VAR_NUMBER) {
+              emsg("Invalid chunk element received: must be a number.");
+              goto theend;
+            }
+          }
+          // Types are ok: setup diff information.
+          n = tv_list_first(i);
+          start_a = TV_LIST_ITEM_TV(n)->vval.v_number;
+          n = TV_LIST_ITEM_NEXT(i, n);
+          count_a = TV_LIST_ITEM_TV(n)->vval.v_number;
+          n = TV_LIST_ITEM_NEXT(i, n);
+          start_b = TV_LIST_ITEM_TV(n)->vval.v_number;
+          n = TV_LIST_ITEM_NEXT(i, n);
+          count_b = TV_LIST_ITEM_TV(n)->vval.v_number;
+          // Counter-mimic extra offsets done by xdiff.
+          if (count_a > 0) {
+            start_a -= 1;
+          }
+          if (count_b > 0) {
+            start_b -= 1;
+          }
+          // For some reason, the a and b expected by xdiff_out
+          // are reversed compared to the chunks produced by vim.diff.
+          xdiff_out(start_b, count_b, start_a, count_a, &diffio->dio_diff);
+        } else {
+          emsg("Invalid chunk received: must contain 4 numbers.");
+          goto theend;
+        }
+      } else {
+        semsg("Invalid chunk type received (%i), expected list (%i).", c_val->v_type, VAR_LIST);
+        goto theend;
+      }
+    }
+  }
+  success = OK;
+theend:
+  tv_list_unref(chunks);
+  return success;
+}
+
 /// Make a diff between files "tmp_orig" and "tmp_new", results in "tmp_diff".
 ///
 /// @param dio
+/// @param idx_orig Buffer number of original file.
+/// @param idx_new Buffer number of new version of the file.
 ///
 /// @return OK or FAIL
-static int diff_file(diffio_T *dio)
+static int diff_file(diffio_T *dio, const int idx_orig, const int idx_new)
 {
   char *tmp_orig = (char *)dio->dio_orig.din_fname;
   char *tmp_new = (char *)dio->dio_new.din_fname;
   char *tmp_diff = (char *)dio->dio_diff.dout_fname;
   if (*p_dex != NUL) {
-    // Use 'diffexpr' to generate the diff file.
-    eval_diff(tmp_orig, tmp_new, tmp_diff);
-    return OK;
+    // Evaluate function with `in_bufnr` and `new_bufnr` arguments.
+    if (diffexpr_internal_prefix()) {
+      // Ignore the first two '->' characters of the option,
+      // they are not part of the function identifier.
+      char_u *fn_name = p_dex + 2;
+      list_T* chunks = eval_diff_fn(idx_orig, idx_new, fn_name);
+      if (chunks == NULL) {
```

Comment:
This check is already done inside `translate_chunks`

Reply:
Very true. Well, I'll rather remove it from `translate_chunks` so I can keep the `pdex + 2` logic located here.